src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs

```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
 				immutable = false;
 			}
 
-			var stuff = GetStuffToCheckForMember(
-				type,
-				typeSyntax,
-				initializer
-			);
+			// Check all initializations of the member for immutability
+			foreach( ExpressionSyntax initializer in initializers ) {
+				var stuff = GetStuffToCheckForMember(
+					type,
+					typeSyntax,
+					initializer
+				);
 
-			if ( stuff == null ) {
-				// null is a signal that there is nothing further that needs to
-				// be checked
-				return immutable;
-			}
+				if( stuff == null ) {
+					// Null is a signal that there is nothing further that needs to
+					// be checked
+					continue;
+				}
 
-			var (typeToCheck, checkKind, getLocation) = stuff.Value;
+				var (typeToCheck, checkKind, getLocation) = stuff.Value;
+
+				if( m_context.IsImmutable(
+					typeToCheck,
+					checkKind,
+					getLocation,
+					out var diagnostic
+				) ) {
+					continue;
+				}
 
-			if( !m_context.IsImmutable( typeToCheck, checkKind, getLocation, out var diagnostic ) ) {
 				diagnosticSink( diagnostic );
 				immutable = false;
 			}
 
 			return immutable;
 		}
 
+
+		/// <summary>
+		/// A readonly field or property may be re-initialized in constructors.
+		/// This method finds all initializations of the field/property.
+		/// </summary>
+		/// <param name="symbol">The field/property to search for</param>
+		/// <param name="initialInitializer">The initialization with the declaration</param>
+		/// <returns>An immutable list of initializations</returns>
+		private ImmutableArray<ExpressionSyntax> GetInitializers(
+			ISymbol symbol,
+			ExpressionSyntax initialInitializer
+		) {
+			List<ExpressionSyntax> initializers = new() {initialInitializer};
+
+			// If the containing symbol isn't a named type (class) then there
+			// will be no constructors

```

Comment:
Members can exist in other INamedTypeSymbols, like records or structs.
What is an example of a field that is not an INamedTypeSymbol?

Reply:
The check is because it is `INamedTypeSymbol` that has the `.Constructors` member.

Reply:
Yeah, so this is similar to the `NullReferenceException`. There are two ways to deal with the cast:

1. Assert that it always exists and suffer an `InvalidCastException` if you're wrong (which will trigger a bug report against the analyzers)
2. Handle the failure to cast somehow

If we do (2) and just return an empty array we are saying that there are no extra writes for non-`INamedTypeSymbol`... so I'm wondering if there is any case that can happen? Does returning `[]` make sense in all those cases? Is there a chance that we miss assignments by doing this, which would be a bug?

If we think `INamedTypeSymbol` is the only option here we should do option (1)